### PR TITLE
feat(1958): hapi ^20.0.0 upgrade. BREAKING CHANGE: New major version

### PR DESCRIPTION
**Pull request by pritamstyz4ever**
## Context

Screwdriver hapi.js dependencies are outdated and not supported anymore.

## Objective

This PR majorly upgrades @hapi/hapi version and all other changes to be compatible with latest @hapi js dependencies.
Also upgrades all screwdriver dependent libraries with a new major version.

## References

https://github.com/screwdriver-cd/screwdriver/issues/1958

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,13 +1,13 @@
 {
   "name": "screwdriver-api",
-  "version": "0.5.0",
+  "version": "1.0.0",
```
Comment (tkyi) [inline package.json:3]: With all of these changes, I think we should bump major version for this repo 👍 
We're currently at v0.5.975

File: plugins/buildClusters/list.js [hunk 2]
```diff
@@ -24,16 +23,18 @@ module.exports = () => ({
...
             const config = {
                 sort: request.query.sort
             };
 
             return buildClusterFactory
                 .list(config)
-                .then(buildClusters => reply(buildClusters.map(c => c.toJson())))
-                .catch(err => reply(boom.boomify(err)));
+                .then(buildClusters => h.response(buildClusters.map(c => c.toJson())))
+                .catch(err => {
```
[the remark this answers is not in the document]
Comment (jithine) [reply plugins/buildClusters/list.js:35]: I see this pattern is repeated all over

File: plugins/auth/login.js [hunk 3]
```diff
@@ -50,99 +49,94 @@ function addGuestRoute(server, config) {
...
-                            if (request.params.web === 'web') {
-                                return reply('<script>window.close();</script>');
-                            }
-
-                            return reply().redirect('/v4/auth/token');
-                        })
-                        .catch(err => reply(boom.boomify(err)))
-                );
+                let user;
+                const model = await userFactory.get({ username, scmContext });
```
[the remark this answers is not in the document]
Comment (pritamstyz4ever) [reply plugins/auth/login.js:98]: Yeah, linter errors.

File: plugins/auth/token.js [hunk 4]
```diff
@@ -22,37 +22,37 @@ module.exports = () => ({
...
-        handler: (request, reply) => {
-            let profile = request.auth.credentials;
-            const { scope, token, username } = profile;
-            const { buildFactory, jobFactory, pipelineFactory } = request.server.app;
-
-            // Check Build ID impersonate
-            if (request.params.buildId) {
-                if (!scope.includes('admin')) {
-                    return reply(boom.forbidden(`User ${username} is not an admin and cannot impersonate`));
+        handler: async (request, h) => {
```
[the remark this answers is not in the document]
Comment (pritamstyz4ever) [reply plugins/auth/token.js:25]: h is hapi response [toolkit`](https://hapi.dev/api/?v=20.0.0#response-toolkit)

File: plugins/buildClusters/update.js [hunk 5]
```diff
@@ -1,14 +1,14 @@
 'use strict';
 
-const boom = require('boom');
-const joi = require('joi');
+const boom = require('@hapi/boom');
 const schema = require('screwdriver-data-schema');
-const nameSchema = joi.reach(schema.models.buildCluster.base, 'name');
+const Joi = require('joi');
```
[the remark this answers is not in the document]
Comment (pritamstyz4ever) [reply plugins/buildClusters/update.js:5]: Yeah, payload validation needs to be a valid joi schema object wrapped with joi.object

File: plugins/banners/get.js [hunk 6]
```diff
@@ -1,19 +1,19 @@
...
 module.exports = () => ({
     method: 'GET',
     path: '/banners/{id}',
-    config: {
+    options: {
         description: 'Get a single banner',
         notes: 'Return a banner record',
         tags: ['api', 'banners'],
-        handler: (request, reply) => {
+        handler: async (request, h) => {
```
[the remark this answers is not in the document]
Comment (pritamstyz4ever) [reply plugins/banners/get.js:16]: standard hapi response [toolkit](https://hapi.dev/api/?v=20.0.0#response-toolkit)

File: plugins/buildClusters/list.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (pritamstyz4ever) [reply plugins/buildClusters/list.js:35]: Its a remanent of removing boom.boomify(err) which is not required and also not valid. We can remove all such throws after converting each handler to async-await syntax.

File: plugins/auth/index.js [hunk 7]
```diff
@@ -22,6 +17,17 @@ const JOI_BOOLEAN = joi
...
     .falsy('false');
 
+/**
+ *
+ * @param {string} decoded
+ * @param {object} request
+ * @param {object} h
+ */
+const validate = async function() {
+    // The _decoded token signature is validated by jwt.veriry so we can return true
```
[the remark this answers is not in the document]
Comment (pritamstyz4ever) [reply plugins/auth/index.js:27]: Yes, its an async fn: https://github.com/dwyl/hapi-auth-jwt2

File: plugins/banners/index.js [hunk 8]
```diff
@@ -13,44 +13,41 @@ const removeRoute = require('./remove');
...
-     * @return {Object}                   Details including the display name and admin status of user
-     */
-    server.expose('screwdriverAdminDetails', (username, scmContext) => {
-        // construct object with defaults to store details
-        const adminDetails = {
-            isAdmin: false
-        };
-
-        if (scmContext) {
-            const { scm } = server.root.app.bannerFactory;
```
[the remark this answers is not in the document]
Comment (pritamstyz4ever) [reply plugins/banners/index.js:31]: server.root is undefined and no longer available and server.app is {} in server.expose.

**Pull request merged**